base/src/main/java/org/aya/core/term/ElimTerm.java

```diff
@@ -10,13 +10,31 @@
 import org.aya.util.distill.DistillerOptions;
 import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Elimination rules.
  *
  * @author ice1000
  */
 public sealed interface ElimTerm extends Term {
+  // ErasedTerm itself and ElimTerm with `of` erased are erased.
+  static @Nullable ErasedTerm checkErased(@NotNull Term term) {
+    if (term instanceof ElimTerm elim) return checkErased(elim.of());
+    if (term instanceof CallTerm.Access elim) return checkErased(elim.of());
+    return term instanceof ErasedTerm erased ? erased : null;
+  }
+  static boolean isErased(@NotNull Term term) {
+    return checkErased(term) != null;
+  }
+  // ErasedTerm with a Prop type might safely appear in IntroTerms.
+  // ErasedTerm with a non-Prop type or ElimTerm with `of` erased are disallowed.
+  static @Nullable ErasedTerm checkErasedNotProp(@NotNull Term term) {
+    if (term instanceof ElimTerm elim) return checkErased(elim.of());
+    if (term instanceof CallTerm.Access elim) return checkErased(elim.of());
+    return term instanceof ErasedTerm erased && !erased.isProp() ? erased : null;
```

Comment:
Are you sure? I guess it might be

```suggestion
  static @Nullable ErasedTerm checkErasedNotProp(@NotNull Term term) {
    if (term instanceof ElimTerm elim) return checkErasedNotProp(elim.of());
    if (term instanceof CallTerm.Access elim) return checkErasedNotProp(elim.of());
    return term instanceof ErasedTerm erased && !erased.isProp() ? erased : null;
```

Reply:
Yes
Consider the following case
```
A : Prop
a : A
a.v : Nat
```
only things like `f a` are allowed